src/Effect/Functor/Naperian.agda

```diff
@@ -0,0 +1,43 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Naperian functor
+--
+-- This file contains definitions of Naperian introduced by Jeremy Gibbons 
+-- in the book APLicative Programming with Naperian Functors.
+-- https://link.springer.com/chapter/10.1007/978-3-662-54434-1_21
+------------------------------------------------------------------------
+
+{-# OPTIONS --cubical-compatible --safe #-}
+
+module Effect.Functor.Naperian where 
+
+open import Effect.Functor using (RawFunctor)
+open import Level using (Level; suc; _⊔_)
+open import Relation.Binary.PropositionalEquality.Core using (_≡_)
+
+private
+  variable
+    b c : Level
+
+
+-- " Functor f is Naperian if there is a type p of ‘positions’ such that fa≃p→a;
+-- then p behaves a little like a logarithm of f
+-- in particular, if f and g are both Naperian, then Log(f×g)≃Logf+Logg and Log(f.g) ≃ Log f × Log g"
+-- APLicative Programming with Naperian Functors. Jeremy Gibbons.
+
+-- RawNaperian contains just the functions, not the proofs
+record RawNaperian {F : Set b → Set c} (d : Level) (RF : RawFunctor F) : Set (suc (b ⊔ d) ⊔ c) where
+  field
+    Log : Set d
+    index : ∀ {A} → F A → (Log → A)
+    tabulate : ∀ {A} → (Log → A) → F A
+
+-- Full Naperian has the coherence conditions too. Propositional version (hard to work with).
```

Comment:
In general, we have had problems in the past with trying to work out how best to tackle equality on `Functor`s (and hence also [formalise the laws](https://github.com/agda/agda-stdlib/issues/496) for them, as well as `Applicative, `Monad, etc.) *in general*, but here, precisely because of the `RawNaperian` structure, we can lift `Setoid` structure `_≈_` on `A` to (the `Pointwise`) one on `Log → A`... and thus induce a `Setoid` structure on `F A` ? (there may well be details to check/fix here!)

... where it's worth noting that `Log` itself should probably be considered only `Propositional`ly, that is, indexing should be considered *intensionally* with respect to the underlying `Set` (perhaps Joyal's theory of species requires more complex `Setoid` (or:other!) indexing, but that's out-of-scope for this PR!)

Reply:
I think this perhaps entails changing the parametrisation of 'full' `Naperian` to build in the `Setoid` structure (otherwise we're back at `Set\omega`), and then back-derive its underlying `RawNaperian` functor. So more thought required!

Reply:
I've done an attempt at this - could you check it out please?